mqtt/mqtt-broker/src/server.rs

```diff
@@ -225,71 +232,124 @@ where
-
-                    let broker_handle = handle.clone();
-                    let span = inner_span.clone();
-                    let authenticator = authenticator.clone();
-                    let make_processor = make_processor.clone();
-                    tokio::spawn(async move {
-                        if let Err(e) =
-                            connection::process(stream, peer, broker_handle, &*authenticator, make_processor)
-                                .instrument(span)
-                                .await
-                        {
-                            warn!(message = "failed to process connection", error =% DetailedErrorValue(&e));
-                        }
-                    });
-                }
-                Either::Left(_) => {
-                    info!(
-                        "accept loop shutdown. no longer accepting connections on {}",
-                        addr
-                    );
-                    break;
+struct Listener {
+    transport: Transport,
+    authenticator: Arc<(dyn Authenticator<Error = Box<dyn StdError>> + Send + Sync)>,
+    ready: Option<Receiver<()>>,
+    broker_handle: BrokerHandle,
+}
+
+impl Listener {
+    fn new<N>(
+        transport: Transport,
+        authenticator: N,
+        broker_handle: BrokerHandle,
+        ready: Option<Receiver<()>>,
+    ) -> Self
+    where
+        N: Authenticator<Error = Box<dyn StdError>> + Send + Sync + 'static,
+    {
+        Self {
+            transport,
+            authenticator: Arc::new(authenticator),
+            ready,
+            broker_handle,
+        }
+    }
+
+    async fn run<F, P>(self, shutdown_signal: F, make_processor: P) -> Result<(), Error>
+    where
+        F: Future<Output = ()> + Unpin,
+        P: MakeIncomingPacketProcessor
+            + MakeOutgoingPacketProcessor
+            + Clone
+            + Send
+            + Sync
+            + 'static,
+    {
+        let Self {
+            transport,
+            authenticator,
+            ready,
+            broker_handle,
+        } = self;
+
+        let addr = transport.addr();
+
+        let span = info_span!("server", listener=%addr);
+        let inner_span = span.clone();
+
+        async move {
+            let ready = match ready {
+                Some(mut ready) => {
+                    info!("Waiting for broker to be ready to serve requests");
+                    // async block required to consume ready and make a future with
+                    // a 'static lifetime
+                    Either::Left(async move { ready.recv().await })
                 }
-                Either::Right((Some(Err(e)), _)) => {
-                    warn!(
-                        message = "accept loop exiting due to an error",
-                        error =% DetailedErrorValue(&e)
-                    );
-                    break;
+                None => Either::Right(future::ready(Ok(()))),
+            };
+            pin_mut!(ready);
```

Comment:
Why is this needed?

Reply:
`future::select` one line below expects some type F that implements both `Future` and `Unpin`. `async` block does not impl `Unpin` so we have to make them pinned

Reply:
and we need `async` block to consume `ready` receiver. borrow checker complains otherwise

Reply:
got it, thx.